File: utils/checkInScore.ts
export interface CheckInScoreState {
  checkInDate: Date;
  physicalStrength: number;
  mentalStrength: number;
  recovered: number;
  confidence: number;
  sleep: number;
  energy: number;
  stress: number;
  soreness: number;
  goal: string;
  readiness: number;
  focus: number;
  excitement: number;
  bodyConnection: number;
  concerns: string;
}

export const createDefaultCheckInScore = (): CheckInScoreState => ({
  checkInDate: new Date(),
  physicalStrength: 3,
  mentalStrength: 3,
  recovered: 3,
  confidence: 3,
  sleep: 3,
  energy: 3,
  stress: 3,
  soreness: 3,
  goal: '',
  readiness: 3,
  focus: 3,
  excitement: 3,
  bodyConnection: 3,
  concerns: '',
});

export const physicalScore = (state: CheckInScoreState): number => {
  const overall = state.physicalStrength + state.recovered + state.energy + (5 - state.soreness) + state.readiness;
  return Math.round(Math.max(0, Math.min(100, (overall / 25) * 100)));
};

export const mentalScore = (state: CheckInScoreState): number => {
  const overall =
    state.mentalStrength +
    state.confidence +
    state.sleep +
    state.stress +
    state.bodyConnection +
    state.focus +
    state.excitement;
  return Math.round(Math.max(0, Math.min(100, (overall / 35) * 100)));
};

export const overallScore = (state: CheckInScoreState): number => {
  const overall =
    state.physicalStrength +
    state.recovered +
    state.energy +
    (5 - state.soreness) +
    state.readiness +
    state.mentalStrength +
    state.confidence +
    state.sleep +
    state.stress +
    state.bodyConnection +
    state.focus +
    state.excitement;
  return Math.round(Math.max(0, Math.min(100, (overall / 60) * 100)));
};

export const hasCompletedCheckIn = (state: CheckInScoreState): boolean => {
  return state.goal.trim().length > 0;
};

File: swift-app/Utils/CheckInScore.swift
//
//  CheckInScore.swift
//  MeetJournal
//
//  Created by Maddisen Mohnsen on 12/25/25.
//

import Foundation
import SwiftUI

@Observable
class CheckInScore {
    var checkInDate: Date = Date()
    var physicalStrength: Int = 3
    var mentalStrength: Int = 3
    var recovered: Int = 3
    var confidence: Int = 3
    var sleep: Int = 3
    var energy: Int = 3
    var stress: Int = 3
    var soreness: Int = 3
    var goal: String = ""
    var readiness: Int = 3
    var focus: Int = 3
    var excitement: Int = 3
    var bodyConnection: Int = 3
    var concerns: String = ""
    
    var physicalScore: Int {
        let overall: Int = physicalStrength + recovered + energy + (5 - soreness) + readiness

        let maxScore = 25.0
        let percent = (Double(overall) / maxScore) * 100.0

        let percentScore = Int(max(0.0, min(100.0, percent)).rounded())
        return percentScore
    }
    
    var mentalScore: Int {
        let overall: Int = mentalStrength + confidence + sleep + stress + bodyConnection + focus + excitement

        let maxScore = 35.0
        let percent = (Double(overall) / maxScore) * 100.0

        let percentScore = Int(max(0.0, min(100.0, percent)).rounded())
        return percentScore
    }
    
    var overallScore: Int {
        let overall: Int = physicalStrength + recovered + energy + (5 - soreness) + readiness + mentalStrength + confidence + sleep + stress + bodyConnection + focus + excitement
        
        let maxScore = 60.0
        let percent = (Double(overall) / maxScore) * 100.0

        let percentScore = Int(max(0.0, min(100.0, percent)).rounded())
        return percentScore
    }
    
    var hasCompletedForm: Bool {
        if goal.isEmpty {
            return false
        }
        
        return true
    }
}
